packages/block-editor/src/utils/show-block-removal-warning.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	Modal,
+	Button,
+	__experimentalHStack as HStack,
+} from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { store as blocksStore } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+
+function isBlockCritical( blockName ) {
+	if ( blockName === 'core/query' || blockName === 'core/post-content' ) {
+		return true;
+	}
+	return false;
+}
+
+export function useBlockRemovalWarning() {
+	const { getBlockName, getBlockOrder } = useSelect( blockEditorStore );
+	const { getBlockType } = useSelect( blocksStore );
+	const { displayRemovalPrompt, removeBlocks } =
+		useDispatch( blockEditorStore );
+
+	function findFirstCriticalBlock( clientIds ) {
+		return clientIds.map( ( clientId ) => {
+			const blockName = getBlockName( clientId );
+			if ( isBlockCritical( blockName ) ) {
+				const blockType = getBlockType( blockName );
+				return blockType?.title || blockName;
+			}
+			const innerBlocks = getBlockOrder( clientId );
+			if ( innerBlocks.length ) {
+				return findFirstCriticalBlock( innerBlocks );
+			}
+			return false;
+		} );
+	}
+
+	return ( clientIds, selectPrevious = true ) => {
+		const firstCriticalBlock = findFirstCriticalBlock( clientIds ).find(
+			( blockName ) => blockName
+		);
+		if ( firstCriticalBlock ) {
+			displayRemovalPrompt( true, {
+				removalFunction: () => {
+					removeBlocks( clientIds, selectPrevious );
+				},
```

Comment:
It's always `removeBlocks`, the only difference is whether `selectPrevious` is true or false. I guess we could just pass `selectPrevious` instead of the whole function.